### ISSUE TITLE
[typescript / rn] How to deal with no type match inside styled() ?

### ISSUE THREAD

**Issue by IljaDaderko** [styled-components/styled-components#1287]
## Version
2.2.3

I'm implementing typescript in our project along with styled components, so far so good, but sometimes cases like this pop up

![example](https://i.imgur.com/kiLz4rA.png)

I assume this is because `styled()` doesn't expect to take in that very specific type (Lineargradient), but I find that some libs export their in a very custom way.

Is there a way to deal with this issue that is common perhaps?

**Comment by philpl** [comment]
cc @styled-components/typers

**Comment by Igorbek** [comment]
@IljaDaderko could you please give a full repro of the issue please? It seems to be the same as #1281 but I could not reproduce it. Maybe I were to reproduce your code.

**Comment by IljaDaderko** [comment]
@Igorbek project is private, but there isn't much to the while where error occurs:

```
import * as React from "react";
import styled from "styled-components/native";
import LinearGradient from "react-native-linear-gradient";
import { purple, lightPurple } from "../style/color";

// -- types ----------------------------------------------------------------- //
export interface Props {
  children: React.ReactChild | React.ReactChild[];
}

// -- styling --------------------------------------------------------------- //
const Card = styled(LinearGradient)`
  width: 100%;
  height: 60px;
  border-radius: 4px;
  shadow-opacity: 0.05;
  shadow-color: #000000;
  shadow-offset: 0px 8px;
  shadow-radius: 7px;
  rc: 20px;
  padding-right: 20px;
  padding-top: 10px;
  padding-bottom: 10px;
  margin-top: 10px;
  margin-bottom: 10px;
  elevation: 2;
  flex-direction: row;
`;

class GradientBlock extends React.Component<Props, {}> {
  // -- render -------------------------------------------------------------- //
  render() {
    const { children } = this.props;
    return (
      <Card colors={[lightPurple, purple]} start={{ x: 0.0, y: 0.5 }} end={{ x: 1.0, y: 0.5 }}>
        {children}
      </Card>
    );
  }
}

export default GradientBlock;
```

here `import { purple, lightPurple } from "../style/color";` are just hex strings like `#000000`

**Comment by Igorbek** [comment]
Please make sure you use latest `@types/react`, `@types/react-dom`, `@types/react-native`

**Comment by jbw** [comment]
This error is the same one I'm getting. https://travis-ci.org/jbw/tyble/jobs/296940825. This project shouldn't build for you if you want something reproducible.

**Comment by IljaDaderko** [comment]
Confirmed, this is on latest typings

**Comment by jbw** [comment]
@IljaDaderko you could cast your component to any `styled(Lineargradient as any)` so you don't have to remove the strict flag for your project.

**Comment by Igorbek** [comment]
@IljaDaderko I tried to reproduce your example and the only issue I found was in the typings of `react-native-linear-gradient`. The typings from `@types/react-native-linear-gradient` declare this:

```ts
declare class LinearGradient extends React.Component<LinearGradientProps> {
    constructor(props?: LinearGradientProps);
}
```
However, `@types/react` declares:
```ts
interface ComponentClass<P = {}> {
    new (props?: P, context?: any): Component<P, ComponentState>;
    propTypes?: ValidationMap<P>;
    contextTypes?: ValidationMap<any>;
    childContextTypes?: ValidationMap<any>;
    defaultProps?: Partial<P>;
    displayName?: string;
}
```

since `ComponentClass` requires constructor to **optionally** accept `props`, and, on other hand, `LinearGradient` requires it (not optional), it makes it incompatible with `ComponentClass`.
It needs to be fixed either in `@types/react-native-linear-gradient` or in `@types/react`. It seems there's [an open PR](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/20987) in `@types/react` that makes this constructor parameter not optional. Likely, it'd be merged soon and fix the issue.
